<a href="https://colab.research.google.com/github/hannahTai/tibame-python/blob/main/python20260829.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

- 快捷鍵
    - 往後縮排: Tab
    - 往前縮排: Shift + Tab

- 完全平方日-想法1
    - 日期往後增加, 字串->數字->開根號->是否整數
    - 普通迴圈

- 完全平方日-想法2
    - 4499 x 4499 -> 20241001 -> is date
    - 4500 x 4500 -> 20250000 -> not date
    - 4501 x 4501 -> 20259001 -> not date

- 模組化、定義函式
- 髒東西: 執行階段>>重新啟動工作階段並執行所有儲存格

In [72]:
# 定義方法 def

def add(a, b):
    c = a + b
    return c

print(add(3, 5))
print(add(4, 6))

8
10


In [73]:
# 閏年方法

def is_leap_year(year):
    if year % 400 == 0:
        return True
    elif year % 100 == 0:
        return False
    elif year % 4 == 0:
        return True
    else:
        return False

print(is_leap_year(2000))
print(is_leap_year(2006))

True
False


In [74]:
# / 相除
# // 相除取整數(取商)(減多少次)
# % (mod) 相除取餘數(取餘數)(減完剩多少)
# 7 / 2 -> 7 -2 -2 -2 ... 1
# ** 0.5 開根號

d = 20241001
day = d % 100
remain = d // 100
month = remain % 100
year = remain // 100
print(year, month, day)


2024 10 1


In [75]:
# 檢查有效日期(My)

def is_valid_date_my(year, month, day):

    month_day = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]

    if is_leap_year(year):
        month_day[1] = 29

    if year == 0:
        return False
    if month == 0 or month > 12:
        return False
    if day == 0 or day > 31:
        return False
    if day > month_day[month -1]:
        return False

    return True

print(is_valid_date_my(2010, 2, 29))
print(is_valid_date_my(2000, 2, 29))
print(is_valid_date_my(2004, 2, 29))
print(is_valid_date_my(1900, 2, 29))


False
True
True
False


In [76]:
# 檢查有效日期(Teacher)

def is_date_valid(d):
    day = d % 100
    remain = d // 100
    month = remain % 100
    year = remain // 100

    if year >= 1:
        if 1 <= month <= 12:
            day_per_month = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
            day_max = day_per_month[month - 1]
            if month == 2 and is_leap_year(year):
                day_max = 29
            if 1 <= day <= day_max:
                return True
            else:
                # day invalid
                return False
        else:
            # month invalid"
            return False
    else:
        # year invalid
        return False

print(is_date_valid(20100229))
print(is_date_valid(20000229))
print(is_date_valid(20040229))
print(is_date_valid(19000229))


False
True
True
False


In [77]:
# 迴圈1: 固定次數
for i in range(3):
    print(i)

# 迴圈2: 固定次數
i = 0
while i < 3:
    print(i)
    i = i + 1

# 迴圈3: 不固定次數
while True:
    break

0
1
2
0
1
2


In [78]:
# 10個完全平方日-想法二(My)

# 20241001: 1
n = 4499
count = 0
while True:
    d = n * n
    if is_date_valid(d):
        count = count + 1
        print(count, ":", n, ",", d)
    if count == 10:
        break
    n = n + 1

1 : 4499 , 20241001
2 : 4552 , 20720704
3 : 4575 , 20930625
4 : 4599 , 21150801
5 : 4625 , 21390625
6 : 4639 , 21520321
7 : 4652 , 21641104
8 : 4653 , 21650409
9 : 4668 , 21790224
10 : 4682 , 21921124


In [79]:
# 10個完全平方日-想法二(Teacher)

n = 4499
count = 0
while True:
    d = n * n
    if is_date_valid(d):
        print(d)
        count = count + 1
    if count == 10:
        break
    n = n + 1

20241001
20720704
20930625
21150801
21390625
21520321
21641104
21650409
21790224
21921124


In [80]:
# 字串專屬功能

s = "hello" * 5
print(s)

# replace (需要=回去!!!)
s = s.replace("hello", "bye")
print(s)

hellohellohellohellohello
byebyebyebyebye


In [81]:
# list專屬功能

l = [20, 30, 40]
print(l)

# append (不用=回去!!!)
l.append(99)
print(l)

[20, 30, 40]
[20, 30, 40, 99]


In [82]:
# 10個完全平方日方法-想法二(My)

def get_perfect_square_date(l, n, size):
    l = []
    count = 0
    while True:
        d = n * n
        if is_date_valid(d):
            count = count + 1
            l.append(d)
        if count == size:
            break
        n = n + 1
    return l

print(get_perfect_square_date(l, 4499, 10))

[20241001, 20720704, 20930625, 21150801, 21390625, 21520321, 21641104, 21650409, 21790224, 21921124]


In [83]:
# 10個完全平方日方法-想法二(Teacher)

def method_2(n=4499, max_count=10):
    result = []
    count = 0
    while True:
        d = n * n
        if is_date_valid(d):
            result.append(d)
            count = count + 1
        if count == max_count:
            break
        n = n + 1
    return result

print(method_2())
print(method_2(max_count=5))
print(method_2(4500))

[20241001, 20720704, 20930625, 21150801, 21390625, 21520321, 21641104, 21650409, 21790224, 21921124]
[20241001, 20720704, 20930625, 21150801, 21390625]
[20720704, 20930625, 21150801, 21390625, 21520321, 21641104, 21650409, 21790224, 21921124, 22071204]


- datetime: 處理日期跟時間
    - strptime: 字串 -> 日期 (parse, 字串轉成 datetime 物件)
    - strftime: 日期 -> 字串 (format, date/datetime 物件轉成字串)
    - datetime: 創建日期
    - timedelta: 日期調整
- time: 時間

In [84]:
import datetime as dt

# strptime: str parse to time
s = "2026-08-29"
d = dt.datetime.strptime(s, "%Y-%m-%d")
print(d)

# strftime: str format from time
s = d.strftime("%m/%d %a")
print(s)

# new datetime (if date not exist, throw exception)
# timedelta (add datetime diff)
d = dt.datetime(2000, 2, 29)
print(d)
delta = dt.timedelta(days=1, weeks=1)
print(d + delta)

2026-08-29 00:00:00
08/29 Sat
2000-02-29 00:00:00
2000-03-08 00:00:00


- 浮點數誤差:
    - 十進位->二進位: 本身就會有無限小數/有限小數
    - 1/3 -> 0.33333333333333333....(十進位) -> 0.1(三進位)
    - 0.33333333333333333.... * 3 -> 0.99999999999999999.... -> 1.0
    - 小數本身就不精確
    - 小數相等 -概念是-> 兩個小數差不多



In [85]:
# 小數
f = 3.2
print(f.is_integer()) # False

f = 3.0
print(f.is_integer()) # True

f = 3.000000000001
print(f.is_integer()) # False

f = 3.000000000000000000000000000000000000000001
print(f.is_integer()) # True

f = 3.000000000000001
print(f.is_integer()) # False

f = 3.0000000000000001 # True
print(f.is_integer())

False
True
False
True
False
True


In [86]:
# 10個完全平方日方法-想法一(My)

def get_perfect_square_date(d=20241001, max_count=10):
    result = []
    count = 0
    while True:
        value = d ** 0.5
        if is_date_valid(d) and value.is_integer():
            count = count + 1
            result.append(d)
        if count == max_count:
            break
        d = d + 1
    return result

print(get_perfect_square_date())
print(get_perfect_square_date(max_count=3))
print(get_perfect_square_date(d=20000101,max_count=3))

[20241001, 20720704, 20930625, 21150801, 21390625, 21520321, 21641104, 21650409, 21790224, 21921124]
[20241001, 20720704, 20930625]
[20151121, 20241001, 20720704]


In [96]:
# 10個完全平方日方法-想法一(Teacher)

# import datetime as dt

def method_1(y=2024, m=10, d=1, max_count=10):
    d = dt.datetime(y, m, d)
    delta = dt.timedelta(days=1)
    count = 0
    result = []
    while True:
        d_int = int(d.strftime("%Y%m%d"))
        d_int_root = d_int ** 0.5
        if d_int_root.is_integer():
            result.append(d_int)
            count = count + 1
        if count == max_count:
            break
        d = d + delta
    return result

print(method_1())
print(method_1(max_count=5))
print(method_1(d=2))

[20241001, 20720704, 20930625, 21150801, 21390625, 21520321, 21641104, 21650409, 21790224, 21921124]
[20241001, 20720704, 20930625, 21150801, 21390625]
[20720704, 20930625, 21150801, 21390625, 21520321, 21641104, 21650409, 21790224, 21921124, 22071204]
